<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>A guide to the single-cell epigenomics analysis - 4&nbsp; Dimension reduction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./file_format.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-3C1W433VY8"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-3C1W433VY8', { 'anonymize_ip': true});
</script>
<script async="" src="https://hypothes.is/embed.js"></script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Dimension reduction</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">A guide to the single-cell epigenomics analysis</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/kaizhang/epigenomics-analysis/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Download" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-download"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./A-guide-to-the-single-cell-epigenomics-analysis.pdf">
            <i class="bi bi-bi-file-pdf pe-1"></i>
          Download PDF
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./A-guide-to-the-single-cell-epigenomics-analysis.epub">
            <i class="bi bi-bi-journal pe-1"></i>
          Download ePub
          </a>
        </li>
    </ul>
    <a href="" title="Share" id="sidebar-tool-dropdown-1" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-1">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.linkedin.com/sharing/share-offsite/?url=|url|">
            <i class="bi bi-bi-linkedin pe-1"></i>
          LinkedIn
          </a>
        </li>
    </ul>
  <a href="" class="quarto-color-scheme-toggle sidebar-tool" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">What is epigenomics?</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./anndata.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">AnnData – Annotated Data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_format.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Input data format</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dim_reduct.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Dimension reduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#preprocessing" id="toc-preprocessing" class="nav-link active" data-scroll-target="#preprocessing"><span class="toc-section-number">4.1</span>  Preprocessing</a></li>
  <li><a href="#spectral-embedding" id="toc-spectral-embedding" class="nav-link" data-scroll-target="#spectral-embedding"><span class="toc-section-number">4.2</span>  Spectral embedding</a></li>
  <li><a href="#matrix-free-spectral-embedding-with-cosine-similarity" id="toc-matrix-free-spectral-embedding-with-cosine-similarity" class="nav-link" data-scroll-target="#matrix-free-spectral-embedding-with-cosine-similarity"><span class="toc-section-number">4.3</span>  Matrix-free spectral embedding with cosine similarity</a></li>
  <li><a href="#nyström-method" id="toc-nyström-method" class="nav-link" data-scroll-target="#nyström-method"><span class="toc-section-number">4.4</span>  Nyström method</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/kaizhang/epigenomics-analysis/edit/main/dim_reduct.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/kaizhang/epigenomics-analysis/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Dimension reduction</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Single-cell ATAC-seq (scATAC-seq) produces large and highly sparse cell by feature count matrix. Working directly with such a large matrix is very inconvinent and computational intensive. Therefore typically, we need to reduce the dimensionality of the count matrix before any downstream analysis. Most of the counts in this matrix are very small. For example, ~50% of the counts are 1 in deeply sequenced scATAC-seq data. As a result, many methods treat the count matrix as a binary matrix.</p>
<p>Different from most existing approaches, the dimension reduction method used in SnapATAC2 is a pairwise-similarity based method, which requires defining and computing similarity between each pair of cells in the data. This method was first proposed in <span class="citation" data-cites="Fang.Ren.2021">(<a href="references.html#ref-Fang.Ren.2021" role="doc-biblioref">Fang et al. 2021</a>)</span>, the version 1 of SnapATAC, and was called “diffusion map”. In SnapATAC2, we reformulate this approach as spectral embedding, <em>a.k.a.</em>, Laplacian eigenmaps.</p>
<section id="preprocessing" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="preprocessing"><span class="header-section-number">4.1</span> Preprocessing</h2>
<p>We preprocess the matrix by the Inverse Document Frequency (IDF) weighting. In the context of scATAC-seq, the IDF is defined as:</p>
</section>
<section id="spectral-embedding" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="spectral-embedding"><span class="header-section-number">4.2</span> Spectral embedding</h2>
<p>Assuming the <span class="math inline">\(n \times p\)</span> cell by feature count matrix <span class="math inline">\(C\)</span> has been preprocessed, we first compute the <span class="math inline">\(n \times n\)</span> pairwise similarity matrix <span class="math inline">\(W\)</span> such that <span class="math inline">\(W_{ij} = \delta(C_{i*}, C_{j*})\)</span>, where <span class="math inline">\(\delta: \mathbb{R}^p \times \mathbb{R}^p \rightarrow \mathbb{R}\)</span> is the function defines the similarity between any two cells. Typical choices of <span class="math inline">\(\delta\)</span> include the jaccard index and the cosine similarity.</p>
<p>We then compute the symmetric normalized graph Laplacian <span class="math inline">\(L_{sym} = I - D^{-1/2} W D^{-1/2}\)</span>, where <span class="math inline">\(I\)</span> is the identity matrix and <span class="math inline">\(D = diag(W1)\)</span>.</p>
<p>The bottom eigenvectors of <span class="math inline">\(L_{sym}\)</span> are selected as the lower dimensional embedding. The corresponding eigenvectors can be computed alternatively as the top eigenvectors of the similarly normalized weight matrix:</p>
<p><span class="math inline">\(\tilde{W} = D^{-1/2} W D^{-1/2}\)</span>,</p>
</section>
<section id="matrix-free-spectral-embedding-with-cosine-similarity" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="matrix-free-spectral-embedding-with-cosine-similarity"><span class="header-section-number">4.3</span> Matrix-free spectral embedding with cosine similarity</h2>
<p>When using the cosine similarity, we can avoid computing the full similarity matrix.</p>
<p>The cosine similarity between two vectors A and B is defined as:</p>
<p><span class="math display">\[S_c(A, B) = \frac{A \cdot B}{||A|| ||B||}\]</span></p>
<p>First we rescale the non-negative count matrix <span class="math inline">\(C\)</span> to <span class="math inline">\(X\)</span> such that the rows of <span class="math inline">\(X\)</span> have unit <span class="math inline">\(L_2\)</span> norm.</p>
<p>The cosine similarity matrix is then defined as,</p>
<p><span class="math display">\[W = XX^T - I\]</span></p>
<p>Note that we set the diagonal of <span class="math inline">\(W\)</span> to zeros by subtracting the identity matrix. This is necessary because our benchmark result show that it generally improves the quality of the embedding.</p>
<p>The degree matrix can be computed as,</p>
<p><span class="math display">\[D = diag((X X^T - I) \mathbf{1}) = diag(X(X^T \mathbf{1}) - \mathbf{1})\]</span></p>
<p>and,</p>
<p><span class="math display">\[\tilde{W} = D^{-1/2} XX^T D^{-1/2} - D^{-1} = \tilde{X}\tilde{X}^T - D^{-1}\]</span></p>
<p>where <span class="math inline">\(\tilde{X} = D^{-1/2} X\)</span>.</p>
<p>Note that <span class="math inline">\(\tilde{X}\)</span> has the same size as <span class="math inline">\(X\)</span>, and if X is sparse, <span class="math inline">\(\tilde{X}\)</span> preserves the sparsity pattern of <span class="math inline">\(X\)</span>.</p>
<p>We remark that this problem would be easier if we ignore the <span class="math inline">\(D^{-1}\)</span> term, because the eigenvectors of <span class="math inline">\(\tilde{X}\tilde{X}^T\)</span> can be computed from the Singular Vector Decomposition (SVD) of <span class="math inline">\(\tilde{X}\)</span>. With the presence of the <span class="math inline">\(D^{-1}\)</span> term, we resort to the Lanczos algorithm to compute the top eigenvectors of <span class="math inline">\(\tilde{W}\)</span> without ever computing <span class="math inline">\(\tilde{W}\)</span>. Each iteration in the Lanczos algorithm requires computing the matrix-vector product between <span class="math inline">\(\tilde{W}\)</span> and <span class="math inline">\(\mathbf{v}\)</span>,</p>
<p><span class="math display">\[\tilde{W} \mathbf{v} = \tilde{X} (\tilde{X}^T \mathbf{v}) - D^{-1} \mathbf{v}\]</span></p>
<p>Using the specific order of operations shown in the formula above, we can reduce the computational cost of the matrix-vector product to <span class="math inline">\(2z + n\)</span>, where <span class="math inline">\(n\)</span> is the number of rows in <span class="math inline">\(X\)</span> and <span class="math inline">\(z\)</span> is the number of non-zero elements in <span class="math inline">\(X\)</span>.</p>
<p>As a comparision, performing this operation on the full similarity matrix will need <span class="math inline">\(n^2\)</span> computations. Note that <span class="math inline">\(z ≪ n^2\)</span> for most scATAC-seq data. Computing the full similarity matrix additionally requires <span class="math inline">\(n^3\)</span> computations using the naive algorithm, which is prohibitively expensive for large datasets. Therefore, the matrix-free method is much faster and more memory efficient.</p>
</section>
<section id="nyström-method" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="nyström-method"><span class="header-section-number">4.4</span> Nyström method</h2>
<p>The matrix-free method described above is very fast and memory efficient. However, for massive datasets with hundreds of millions of cells, storing the cell by peak count matrix <span class="math inline">\(C\)</span> may already be a challenge. In this section, we describe an on-line embedding method that can applied to virtually arbitrary large datasets. The key idea here is to use the Nystrom method to perform a low-rank approximation of the full similarity matrix.</p>
<p>We will be focusing on generating an approximation <span class="math inline">\(\tilde{W}\)</span> of <span class="math inline">\(W\)</span> based on a sample of <span class="math inline">\(l ≪ n\)</span> of its columns.</p>
<p>Suppose <span class="math inline">\(W = \begin{bmatrix} A &amp; B \\ B^T &amp; C \end{bmatrix}\)</span> and columns <span class="math inline">\(\begin{bmatrix} A \\ B^T \end{bmatrix}\)</span> are our samples. We first perform eigendecomposition on <span class="math inline">\(A = U \Lambda U^T\)</span>. The nystrom method approximates the eigenvectors of matrix <span class="math inline">\(W\)</span> by <span class="math inline">\(\tilde{U} = \begin{bmatrix} U \\ B^T U \Lambda^{-1} \end{bmatrix}\)</span>.</p>
<p>We can then compute <span class="math inline">\(\tilde{W}\)</span>:</p>
<p><span class="math display">\[
\begin{aligned}
\tilde{W} &amp;= \tilde{U} \Lambda \tilde{U}^T \\
          &amp;= \begin{bmatrix} U \\ B^T U \Lambda^{-1} \end{bmatrix}
             \Lambda
             \begin{bmatrix} U^T &amp; \Lambda^{-1}U^TB \end{bmatrix} \\
          &amp;= \begin{bmatrix}
               U \Lambda U^T &amp; U \Lambda \Lambda^{-1} U^T B \\
               B^T U \Lambda^{-1} \Lambda U^T &amp; B^T U \Lambda^{-1} \Lambda \Lambda^{-1} U^T B
             \end{bmatrix} \\
          &amp;= \begin{bmatrix} A &amp; B \\ B^T &amp; B^T U \Lambda^{-1} U^T B \end{bmatrix}
\end{aligned}
\]</span></p>
<p>In practice, <span class="math inline">\(\tilde{W}\)</span> does not need to be computed. Instead, it is used implicitly to estimate the degree normalization vector:</p>
<p><span class="math display">\[
\tilde{d} = \tilde{W}\mathbf{1} = \begin{bmatrix}A\mathbf{1} + B\mathbf{1} \\ B^T \mathbf{1} + B^T A^{-1} B\mathbf{1}
\end{bmatrix}
\]</span></p>
<p>This approach requires computing the inverse of <span class="math inline">\(A\)</span>, which is expensive when <span class="math inline">\(A\)</span> is large. Here we use an algorithm reported in XXX to approximate the degree matrix.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography" style="display: none">
<div id="ref-Fang.Ren.2021" class="csl-entry" role="doc-biblioentry">
Fang, Rongxin, Sebastian Preissl, Yang Li, Xiaomeng Hou, Jacinta Lucero, Xinxin Wang, Amir Motamedi, et al. 2021. <span>“<span class="nocase">Comprehensive analysis of single cell ATAC-seq data with SnapATAC</span>.”</span> <em>Nature Communications</em> 12 (1): 1337. <a href="https://doi.org/10.1038/s41467-021-21583-9">https://doi.org/10.1038/s41467-021-21583-9</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./file_format.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Input data format</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>